txstratum/cli.py

```diff
@@ -51,110 +53,154 @@ def create_parser() -> ArgumentParser:
-        from txstratum.prometheus import HttpPrometheusExporter
-        http_metrics = HttpPrometheusExporter(manager, pubsub, args.prometheus_port)
-        http_metrics.start()
-        logger.info('Prometheus metrics server running at 0.0.0.0:{}...'.format(args.prometheus_port))
-
-    if args.ban_addrs or args.ban_tx_ids:
-        tx_filters.append(FileFilter.load_from_files(args.ban_tx_ids, args.ban_addrs))
-
-    if args.toi_url or args.toi_apikey:
-        if not (args.toi_url and args.toi_apikey):
-            raise ValueError("Should pass both toi_url and toi_apikey")
-        toiclient = TOIAsyncClient(args.toi_url, args.toi_apikey)
-        tx_filters.append(TOIFilter(toiclient, block=args.toi_fail_block))
-
-    api_app = App(manager, max_tx_weight=args.max_tx_weight, max_timestamp_delta=args.max_timestamp_delta,
-                  tx_timeout=args.tx_timeout, fix_invalid_timestamp=args.fix_invalid_timestamp,
-                  only_standard_script=not args.allow_non_standard_script, tx_filters=tx_filters)
-    logger.info('API Configuration', max_tx_weight=api_app.max_tx_weight, tx_timeout=api_app.tx_timeout,
-                max_timestamp_delta=api_app.max_timestamp_delta, fix_invalid_timestamp=api_app.fix_invalid_timestamp,
-                only_standard_script=api_app.only_standard_script, tx_filters=tx_filters)
-
-    web_runner = web.AppRunner(api_app.app, logger=logger)
-    loop.run_until_complete(web_runner.setup())
-    site = web.TCPSite(web_runner, '0.0.0.0', args.api_port)
-    loop.run_until_complete(site.start())
-
-    try:
-        logger.info('Stratum Server running at 0.0.0.0:{}...'.format(args.stratum_port))
-        logger.info('TxMining API running at 0.0.0.0:{}...'.format(args.api_port))
-        if args.testnet:
-            logger.info('Running with testnet config file')
-        loop.run_forever()
-    except KeyboardInterrupt:
-        logger.info('Stopping...')
+class RunService:
+    """This is the main class of the service. It starts everything up."""
+
+    manager: TxMiningManager
+    loop: AbstractEventLoop
+
+    async def graceful_shutdown(self) -> None:
+        """Gracefully shutdown the service."""
+        logger.info('Gracefully shutting down...')
+
+        while len(self.manager.tx_queue) > 0:
+            logger.info('Waiting for pending txs to finish...', txs_left=len(self.manager.tx_queue))
+            await asyncio.sleep(2)
+
+        self.manager.shutdown()
+
+    def sigterm_handler(self) -> None:
+        """Handle SIGTERM signal."""
+        logger.info('SIGTERM received.')
+
+        self.loop.create_task(self.graceful_shutdown())
+
+    def register_signal_handlers(self) -> None:
+        """Register signal handlers."""
+        import signal
+
+        logger.info('Registering signal handlers...')
+
+        sigterm = getattr(signal, 'SIGTERM', None)
+        if sigterm is not None:
+            self.loop.add_signal_handler(sigterm, self.sigterm_handler)
```

Comment:
Does this play well with SIGINT signal (or KeyboardInterrupt exception)? I think shutdown by Ctrl+C is useful too, so it might make sense to handle these too.

Reply:
I can test adding a handler for SIGINT as well.

In Kubernetes it won't be possible to stop the process with Ctrl + C. The normal flow when a pod is deleted is that a SIGTERM is sent to the main process in the container, then a SIGKILL some seconds later. That's why I just added the SIGTERM handler.

But I guess we could have it handle SIGINT as well in cases where this is run in other kinds of infrastructure.

I will test this and make a new commit if it works well.

Reply:
I meant for development. Mostly because I think a similar approach will be useful for `hathor-core`, which already gracefully shutsdown on Ctrl+C but not on SIGTERM. (a Ctrl+C from the OS POV is just a SIGINT, but Python will "convert" it to a KeyboardInterrupt exception, I'm not too familiar with the specific mechanics of how this works though).

Reply:
I've implemented handling of SIGINT in https://github.com/HathorNetwork/tx-mining-service/pull/53/commits/6f3d75d339936025f8e922430f629749b360f552

Python stopped converting it into a exception when I started handling it.

Reply:
Great, thanks!